### DISCUSSION TITLE
Add generics to the container (and everywhere)

### DISCUSSION THREAD

**Discussion by ostark** [craftcms/cms#10124] [Ideas] [answered]
PHPStorm, phpstan and psalm added support for the [class-string generic in doc blocks](https://phpstan.org/writing-php-code/phpdoc-types#class-string) earlier this year.

Not sure if this should belong to the framework (Yii) or if you are willing to overwrite some container methods to achieve this:

<img width="784" alt="Bildschirmfoto 2021-11-19 um 14 37 59" src="https://user-images.githubusercontent.com/1941743/142632201-a7e607ea-7acd-4694-9f4f-a3593fedfd5e.png">

The type declaration in this example could look like this:

```php
    /**
     * @template TClassName
     * @param string|array|callable|class-string<TClassName> $type
     * @param array $params the constructor parameters
     * @return TClassName
     */
    public static function createObject($type, array $params = []): object
    {
        return parent::createObject($type, $params);
    }
```

**Comment by brandonkelly** [comment, marked as answer]
Yeah I tried to get this added to Yii directly (https://github.com/yiisoft/yii2/pull/18976) but they didn’t accept it.  Just went ahead and added to Craft directly (efb6784f6faee98bce325b3aedc2f1eaabd3dd6a).

**Reply by ostark** [reply]
Ah, cool - didn't know this was possible
array{class: class-string<T>

**Reply by brandonkelly** [reply]
Yeah PhpStorm doesn’t know what to do with that (yet) but it’s supported by PHP Stan.
